### ISSUE TITLE
Trouble rendering Contentful embedded asset in rich text field 

### ISSUE THREAD

**Issue by satcreative** [gatsbyjs/gatsby#21620]
Using Gatsby & Contentful I'm trying to render an embedded asset in a rich text field. I've seen examples that call the asset url property (along with other properties such as the title) that is contained in node.data.target.fields. However the fields data object isn't available when I query the rich text json data. Below is my query and data output:

```
query {
  allContentfulBlogPost {
    edges {
      node {
        postContent {
          json
        }
      }
    }
  }
}
```

```
"postContent": {
  "json": {
    "data": {},
    "content": [
      {
        "data": {
          "target": {
            "sys": {
              "id": "c1NQ40Elf6OT72FfB6OawHl",
              "type": "Link",
              "linkType": "Asset",
              "contentful_id": "1NQ40Elf6OT72FfB6OawHl"
            }
          }
        },
        "content": [],
        "nodeType": "embedded-asset-block"
      }
    ]
  }
}
```
Do you know why I'm missing the fields data object? Or how I can render the asset using the id value, as this is the only data that I'm being given?

---------------------------------------------------------------

  System:
    OS: macOS 10.15.2
    CPU: (4) x64 Intel(R) Core(TM) i5-4670 CPU @ 3.40GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 12.13.1 - /usr/local/bin/node
    Yarn: 1.17.3 - /usr/local/bin/yarn
    npm: 6.12.1 - /usr/local/bin/npm
  Languages:
    Python: 2.7.13 - /Applications/MAMP/Library/bin/python
  Browsers:
    Chrome: 79.0.3945.130
    Firefox: 72.0.2
    Safari: 13.0.4
  npmPackages:
    gatsby: ^2.18.18 => 2.18.18
    gatsby-image: ^2.2.37 => 2.2.37
    gatsby-plugin-google-analytics: ^2.1.34 => 2.1.34
    gatsby-plugin-htaccess: ^1.4.0 => 1.4.0
    gatsby-plugin-manifest: ^2.2.34 => 2.2.34
    gatsby-plugin-netlify: ^2.1.30 => 2.1.30
    gatsby-plugin-offline: ^2.2.10 => 2.2.10
    gatsby-plugin-page-transitions: ^1.0.8 => 1.0.8
    gatsby-plugin-react-helmet: ^3.1.18 => 3.1.18
    gatsby-plugin-react-helmet-canonical-urls: ^1.4.0 => 1.4.0
    gatsby-plugin-sass: ^2.1.26 => 2.1.26
    gatsby-plugin-sharp: ^2.3.10 => 2.3.10
    gatsby-plugin-sitemap: ^2.2.26 => 2.2.26
    gatsby-remark-images: ^3.1.39 => 3.1.39
    gatsby-remark-lazy-load: ^1.0.2 => 1.0.2
    gatsby-remark-relative-images: ^0.2.3 => 0.2.3
    gatsby-source-contentful: ^2.1.73 => 2.1.73
    gatsby-source-filesystem: ^2.1.43 => 2.1.43
    gatsby-source-instance-name-for-remark: ^0.1.1 => 0.1.1
    gatsby-transformer-remark: ^2.6.45 => 2.6.45
    gatsby-transformer-sharp: ^2.3.9 => 2.3.9

### File contents (if changed)

`gatsby-config.js`: 
```
module.exports = {
  siteMetadata: {
    ...
  },
  plugins: [
	{
      resolve: 'gatsby-plugin-htaccess',
      options: {
        RewriteBase: true,
        https: true,
        www: false,
        host: ...,
      },
    },
	{
	  resolve: `gatsby-plugin-sitemap`,
	  options: {
	      exclude: [`/contact/thanks`, `/privacy-policy`, `/terms-conditions`],
	  }
	},
	{
      resolve: `gatsby-plugin-google-analytics`,
      options: {
        // The property ID; the tracking code won't be generated without it
        trackingId: ...,
        // Defines where to place the tracking script - `true` in the head and `false` in the body
        head: true,
        // Setting this parameter is optional
        anonymize: true,
        // Setting this parameter is also optional
        respectDNT: true,
      },
    },
  	`gatsby-plugin-sharp`,
    `gatsby-transformer-sharp`,
    {
		resolve: `gatsby-transformer-remark`,
		options: {
			plugins: [
				`gatsby-remark-relative-images`,
				{
					resolve: `gatsby-remark-images`,
					options: {
						maxWidth: 900,
						linkImagesToOriginal: false
					},
				},
			]
		},
	},
    `gatsby-plugin-react-helmet`,
    {
		resolve: `gatsby-plugin-react-helmet-canonical-urls`,
		options: {
			siteUrl: `https://satellitecreative.com`,
		},
	},
    `gatsby-plugin-sass`,
    {
		resolve: `gatsby-source-contentful`,
		options: {
			spaceId: process.env.CONTENTFUL_SPACE_ID,
	        accessToken: process.env.CONTENTFUL_ACCESS_TOKEN,
	        downloadLocal: true,
	        host: `preview.contentful.com`,
		},
	},
    {
      resolve: `gatsby-source-filesystem`,
      options: {
	    name: `team`,
        path: `${__dirname}/src/data/team`,
      },
    },
    {
      resolve: `gatsby-source-filesystem`,
      options: {
	    name: `images`,
        path: `${__dirname}/src/images`,
      },
    },
    {
      resolve: `gatsby-plugin-manifest`,
      options: {
        ...
      },
    },
    `gatsby-plugin-netlify`,
    `gatsby-plugin-offline`,
  ],
}

```
`package.json`: 
```
{
  "name": "gatsby-starter-default",
  "private": true,
  "description": "A simple starter to get up and developing quickly with Gatsby",
  "version": "0.1.0",
  "author": "Kyle Mathews <mathews.kyle@gmail.com>",
  "dependencies": {
    "@contentful/rich-text-html-renderer": "^13.4.0",
    "@contentful/rich-text-react-renderer": "^13.4.0",
    "gatsby": "^2.18.18",
    "gatsby-image": "^2.2.37",
    "gatsby-plugin-google-analytics": "^2.1.34",
    "gatsby-plugin-htaccess": "^1.4.0",
    "gatsby-plugin-manifest": "^2.2.34",
    "gatsby-plugin-netlify": "^2.1.30",
    "gatsby-plugin-offline": "^2.2.10",
    "gatsby-plugin-page-transitions": "^1.0.8",
    "gatsby-plugin-react-helmet": "^3.1.18",
    "gatsby-plugin-react-helmet-canonical-urls": "^1.4.0",
    "gatsby-plugin-sass": "^2.1.26",
    "gatsby-plugin-sharp": "^2.3.10",
    "gatsby-plugin-sitemap": "^2.2.26",
    "gatsby-remark-images": "^3.1.39",
    "gatsby-remark-lazy-load": "^1.0.2",
    "gatsby-remark-relative-images": "^0.2.3",
    "gatsby-source-contentful": "^2.1.73",
    "gatsby-source-filesystem": "^2.1.43",
    "gatsby-source-instance-name-for-remark": "^0.1.1",
    "gatsby-transformer-remark": "^2.6.45",
    "gatsby-transformer-sharp": "^2.3.9",
    "lazysizes": "^5.2.0",
    "node-sass": "^4.13.0",
    "prop-types": "^15.7.2",
    "react": "^16.12.0",
    "react-dom": "^16.12.0",
    "react-draggable": "^4.2.0",
    "react-google-recaptcha": "^2.0.1",
    "react-helmet": "^5.2.1",
    "react-hook-form": "^3.29.4",
    "react-spring": "^8.0.27",
    "react-with-gesture": "^4.0.8"
  },
  "devDependencies": {
    "env-cmd": "^10.0.1",
    "prettier": "^1.19.1"
  },
  "keywords": [
    "gatsby"
  ],
  "license": "MIT",
  "scripts": {
    "build": "gatsby build",
    "develop": "env-cmd gatsby develop",
    "dev-m": "env-cmd gatsby develop -H 0.0.0.0 -p 8000",
    "format": "prettier --write \"**/*.{js,jsx,json,md}\"",
    "start": "npm run develop",
    "serve": "gatsby serve",
    "test": "echo \"Write tests! -> https://gatsby.dev/unit-testing \""
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/gatsbyjs/gatsby-starter-default"
  },
  "bugs": {
    "url": "https://github.com/gatsbyjs/gatsby/issues"
  }
}
```
`gatsby-node.js`: 
```
const path = require('path')

module.exports.createPages = async({ graphql, actions }) => {
	const { createPage } = actions
	const clientTemplate = path.resolve(`./src/templates/client.js`)
	const blogTemplate = path.resolve(`./src/templates/blog-list.js`)
	const categoryTemplate = path.resolve(`./src/templates/blog-category-list.js`)
	const postTemplate = path.resolve(`./src/templates/blog-post.js`)
	const res = await graphql(`
		query {
			allContentfulClient ( sort: { order: ASC, fields: slug } ) {
				edges {
					node {
						slug
					}
				}
			}
			allContentfulBlogPost ( sort: { order: DESC, fields: createdAt } ) {
				edges {
					node {
						slug
					}
				}
			}
		}
	`)
	
	const resNewsPosts = await graphql(`
		query {
			allContentfulBlogPost ( 
				filter: { categories: { elemMatch: { slug: { eq: "news" } } } },
				sort: { order: DESC, fields: createdAt } 
			) {
				edges {
					node {
						slug
					}
				}
			}
		}
	`)
	
	const resReviewPosts = await graphql(`
		query {
			allContentfulBlogPost ( 
				filter: { categories: { elemMatch: { slug: { eq: "reviews" } } } },
				sort: { order: DESC, fields: createdAt } 
			) {
				edges {
					node {
						slug
					}
				}
			}
		}
	`)
	
	const resTutorialPosts = await graphql(`
		query {
			allContentfulBlogPost ( 
				filter: { categories: { elemMatch: { slug: { eq: "tutorials" } } } },
				sort: { order: DESC, fields: createdAt } 
			) {
				edges {
					node {
						slug
					}
				}
			}
		}
	`)
	
	const clients = res.data.allContentfulClient.edges
	
	clients.forEach((client, index) => {
		
		const prev = index === 0 ? clients[clients.length - 1].node.slug : clients[index - 1].node.slug
		const next = index === (clients.length - 1) ? clients[0].node.slug : clients[index + 1].node.slug
		
		createPage({
			component: clientTemplate,
			path: `/client/${client.node.slug}/`,
			context: {
				slug: client.node.slug,
				prev: `/client/${prev}`,
				next: `/client/${next}`,
			}
		})
	})
		
	const postsPerPage = 2
	
	const posts = res.data.allContentfulBlogPost.edges
	const numPagesAllPosts = Math.ceil(posts.length / postsPerPage)

	Array.from({ length: numPagesAllPosts }).forEach((_, i) => {
		createPage({
			path: i === 0 ? `/blog/` : `/blog/${i + 1}`,
			component: blogTemplate,
			context: {
				limit: postsPerPage,
				skip: i * postsPerPage,
				numPages: numPagesAllPosts,
				currentPage: i + 1,
				urlPath: `/blog/`
			},
		})
	})
	
	const newsPosts = resNewsPosts.data.allContentfulBlogPost.edges
	const numPagesNewsPosts = Math.ceil(newsPosts.length / postsPerPage)

	Array.from({ length: numPagesNewsPosts }).forEach((_, i) => {
		createPage({
			path: i === 0 ? `/blog/news/` : `/blog/news/${i + 1}`,
			component: categoryTemplate,
			context: {
				filter: 'news',
				limit: postsPerPage,
				skip: i * postsPerPage,
				numPages: numPagesNewsPosts,
				currentPage: i + 1,
				urlPath: `/blog/news/`
			},
		})
	})
	
	const reviewPosts = resReviewPosts.data.allContentfulBlogPost.edges
	const numPagesReviewPosts = Math.ceil(reviewPosts.length / postsPerPage)

	Array.from({ length: numPagesReviewPosts }).forEach((_, i) => {
		createPage({
			path: i === 0 ? `/blog/reviews/` : `/blog/reviews/${i + 1}`,
			component: categoryTemplate,
			context: {
				filter: 'reviews',
				limit: postsPerPage,
				skip: i * postsPerPage,
				numPages: numPagesReviewPosts,
				currentPage: i + 1,
				urlPath: `/blog/reviews/`
			},
		})
	})
	
	const tutorialPosts = resTutorialPosts.data.allContentfulBlogPost.edges
	const numPagesTutorialPosts = Math.ceil(tutorialPosts.length / postsPerPage)

	Array.from({ length: numPagesTutorialPosts }).forEach((_, i) => {
		createPage({
			path: i === 0 ? `/blog/tutorials/` : `/blog/tutorials/${i + 1}`,
			component: categoryTemplate,
			context: {
				filter: 'tutorials',
				limit: postsPerPage,
				skip: i * postsPerPage,
				numPages: numPagesTutorialPosts,
				currentPage: i + 1,
				urlPath: `/blog/tutorials/`
			},
		})
	})

	posts.forEach((post, index) => {
		
		const prev = index === 0 ? posts[posts.length - 1].node.slug : posts[index - 1].node.slug
		const next = index === (posts.length - 1) ? posts[0].node.slug : posts[index + 1].node.slug
		
		createPage({
			component: postTemplate,
			path: `/blog/${post.node.slug}/`,
			context: {
				slug: post.node.slug,
				prev: `/blog/${prev}`,
				next: `/blog/${next}`,
			}
		})
	})
	
}
```
`gatsby-browser.js`: N/A 
`gatsby-ssr.js`: N/A

**Comment by freiksenet** [comment]
Hello!

Have you tried solution for rendering RichText described in [contentful source guide](https://www.gatsbyjs.org/packages/gatsby-source-contentful/#contentful-rich-text)?

Thanks!

**Comment by satcreative** [comment]
Hi @freiksenet. I have, yeah. That example only documents rendering basic text, which I'm able to do without an issue. It doesn't go into embedded assets and the like.

I'm able to query the JSON for text elements, however it's the asset/entry field data that isn't available to me, which makes me think it's a config issue rather than a template/render issue.

Thanks

**Comment by mrmartineau** [comment]
@satcreative could you reply with the query you're using on the page where the rich text is being rendered? My query looks like this:

```js
export const query = graphql`
  query($id: String, $locale: String) {
    contentfulPageArticle(
      contentful_id: { eq: $id }
      node_locale: { eq: $locale }
    ) {
      articleContent {
        json
      }
    }
  }
`
```

I wonder if the response is different when you query using `all*` (`allContentfulBlogPost`) vs `contentfulBlogPost`..

**Comment by satcreative** [comment]
@mrmartineau, I'm using the below:

```
export const query = graphql`
	query ($slug: String!) {
		contentfulBlogPost ( slug: { eq: $slug } ) {
			title
			postContent { 
				json 
			}
		}
	}
`
```

Not too dissimilar to your query. I've tried your exact query in my graphQL playground and am still not seeing the asset URL. It must be something to do with my Contentful setup or the way I'm loading the Contentful data into my site.

Does your Contentful config look similar to mine?

**Comment by mrmartineau** [comment]
@satcreative the only difference I have is that I'm not using the `downloadLocal: true` option

**Comment by satcreative** [comment]
Just resolved it.

It was due to the development script I was running it on. I was using the standard npm run gatsby develop, however I had set up another script a while ago to [access my dev site on my mobile](https://www.gatsbyjs.org/docs/gatsby-cli/#preview-changes-on-other-devices), gatsby develop -H 0.0.0.0 -p 8000. This seems to have done the trick, not really sure why thought, but at least I'm getting the data now. 

Thanks for your help!
